polaris/ocean/ice_shelf/__init__.py

```diff
@@ -0,0 +1,75 @@
+
+
+class IceShelfTask(Task):
+
+    """
+    shared_steps : dict of dict of polaris.Steps
+        The shared steps to include as symlinks
+    """
+    def __init__(self, component, resolution, name, subdir, sshdir=None):
+        if sshdir is None:
+            sshdir = subdir
+        super().__init__(component=component, name=name, subdir=subdir)
+        self.sshdir = sshdir
+        self.component = component
+        self.resolution = resolution
+
+    def _setup_ssh_adjustment_steps(self, init, config, config_filename,
+                                    package=None,
+                                    yaml_filename='ssh_forward.yaml',
+                                    yaml_replacements=None):
+
+        resolution = self.resolution
+        component = self.component
+        indir = self.sshdir
+
+        num_iterations = config.getint('ssh_adjustment', 'iterations')
+        shared_steps: Dict[str, Step] = dict()
+
+        iteration = 0
+        name = f'ssh_forward_{iteration}'
+        ssh_forward = SshForward(
+            component=component, resolution=resolution, indir=indir,
+            mesh=init, init=init, name=name, package=package,
+            yaml_filename=yaml_filename, yaml_replacements=yaml_replacements)
+        ssh_forward.set_shared_config(config, link=config_filename)
+        shared_steps[name] = ssh_forward
+
+        for iteration in range(1, num_iterations):
+            name = f'ssh_adjust_{iteration - 1}'
+            ssh_adjust = SshAdjustment(
+                component=component, resolution=resolution, indir=indir,
+                name=name, init=init, forward=ssh_forward)
+            ssh_adjust.set_shared_config(config, link=config_filename)
+            shared_steps[name] = ssh_adjust
+            name = f'ssh_forward_{iteration}'
+            ssh_forward = SshForward(
+                component=component, resolution=resolution, indir=indir,
+                mesh=init, init=ssh_adjust, name=name, package=package,
+                yaml_filename=yaml_filename,
+                yaml_replacements=yaml_replacements)
+            ssh_forward.set_shared_config(config, link=config_filename)
+            shared_steps[name] = ssh_forward
+
+        iteration = num_iterations
+        name = f'ssh_adjust_{iteration - 1}'
+        ssh_adjust = SshAdjustment(
+            component=component, resolution=resolution, indir=indir,
+            name=name, init=init, forward=ssh_forward)
+        ssh_adjust.set_shared_config(config, link=config_filename)
+        shared_steps[name] = ssh_adjust
+        subdir = f'{indir}/ssh_adjustment/{name}'
+        if subdir in component.steps:
+            ssh_adjust = component.steps[subdir]
+
+        for name, shared_step in shared_steps.items():
+            subdir = f'{indir}/ssh_adjustment/{name}'
+            if subdir in component.steps:
+                shared_step = component.steps[subdir]
+            self.add_step(shared_step, symlink=f'ssh_adjustment/{name}')
```

Comment:
Sure, that makes the code cleaner. I moved it into the for-loop but I still have these lines called always because I found that if I put them inside the if-statement they did not appear in a `polaris list --verbose`. I might be missing something though.

https://github.com/E3SM-Project/polaris/blob/75721a60a063ee4fd5f5adbaff157321920760f2/polaris/ocean/ice_shelf/__init__.py#L38-L42
https://github.com/E3SM-Project/polaris/blob/75721a60a063ee4fd5f5adbaff157321920760f2/polaris/ocean/ice_shelf/__init__.py#L52-L54

Reply:
>  I moved it into the for-loop but I still have these lines called always because I found that if I put them inside the if-statement they did not appear in a polaris list --verbose. I might be missing something though.

I think you want have:
``` python
            subdir = f'{indir}/ssh_adjustment/{name}'
            if subdir in component.steps:
                shared_step = component.steps[subdir]
            else:
                ssh_forward = SshForward(
                    component=component, resolution=resolution, indir=indir,
                    mesh=init, init=current_init, name=name, package=package,
                    yaml_filename=yaml_filename,
                    yaml_replacements=yaml_replacements)
                ssh_forward.set_shared_config(config, link=config_filename)
                shared_step = ssh_forward
```
and similarly for the `SshAdjustment` step.  That way, the steps don't get created and then thrown away.

Reply:
Hmm, I think there's a more fundamental issue.  If `ssh_forward` and `ssh_adjust` are going to be shared steps, they need to both have a unique directory outside of `indir` and a symlink within `indir`.  But the code here seems like it's trying to make a symlink to the same place where the step will be placed.

Reply:
I might be getting confused because `sshdir` gets renamed to `indir`.  Is that a directory within the test case or a shared directory outside of it (or is is meant to be either one)?

Reply:
Okay, I think I understand better now.  So far, there are shared steps and a symlink because `sshdir` is always different from `subdir`.  But I think we want to handle the case where `sshdir == subdir`.  I *think* that would be:
```python
            if indir == self.subdir:
                symlink = None
            else:
                symlink = f'ssh_adjustment/{name}'
            self.add_step(shared_step, symlink=symlink)
```

Reply:
@xylar Thank you so much! I didn't think of that. That did the trick!